### ISSUE TITLE
core library support

### ISSUE THREAD

**Issue by cdepillabout** [purenix-org/purenix#37]
This issue is to track which core libraries have been ported to PureNix.

### TODO

- [ ] [catenable-lists](https://github.com/purescript/catenable-lists)
- [ ] [datetime](https://github.com/purescript/datetime)  (This might be difficult? I don't think Nix has any built-in support for working with dates or times.)
- [ ] [enums](https://github.com/purescript/enums)
- [ ] [filterable](https://github.com/purescript/filterable)
- [ ] [foreign](https://github.com/purescript/foreign)
- [ ] [foreign-object](https://github.com/purescript/foreign-object)
- [ ] [free](https://github.com/purescript/free)
- [ ] [functions](https://github.com/purescript/functions)
- [ ] [graphs](https://github.com/purescript/graphs)
- [ ] [integers](https://github.com/purescript/integers)
- [ ] [lazy](https://github.com/purescript/lazy)  (This isn't technically necessary, because PureNix is lazy.  But it may be nice to have for library compatibility.  Also, we probably want s
omething like Haskell's [deepseq](https://hackage.haskell.org/package/deepseq) instead.)
- [ ] [lcg](https://github.com/purescript/lcg)
- [ ] [lists](https://github.com/purescript/lists)
- [ ] [math](https://github.com/purescript/math)
- [ ] [numbers](https://github.com/purescript/numbers)
- [ ] [ordered-collections](https://github.com/purescript/ordered-collections)
- [ ] [psci-support](https://github.com/purescript/psci-support)  (I'm not sure what PureNix needs to do to get PSCi working.  This library may or may not be needed.)
- [ ] [record](https://github.com/purescript/record)
- [ ] [semirings](https://github.com/purescript/semirings)
- [ ] [strings](https://github.com/purescript/strings)  (This would be hard to port as-is, since Nix doesn't have any built-in support for UTF8 or Unicode.  However, @-sternenseemann has wri
tten a proof-of-concept [UTF8 library](https://code.tvl.fyi/tree/users/sterni/nix/utf8/default.nix?id=750ef6c6934209262cf6dfc00139ca32d08f442d) for Nix that we could likely draw inspiration 
from.)
- [ ] [typelevel-prelude](https://github.com/purescript/typelevel-prelude)
- [ ] [validation](https://github.com/purescript/validation)

### WON'T DO / CAN'T DO

-   [assert](https://github.com/purescript/assert)

    Nix does have an `assert` statement, but I'm not sure that would be able to
    be used with a similar API to `purescript-assert`.

-   [console](https://github.com/purescript/console)

    Nix doesn't have any way of writing arbitrary text to the console.  There
    is the `builtins.trace` function, but that could be wrapped up in it's own
    library.

-   [docs-search](https://github.com/purescript/docs-search)

    I'm not sure what this is or how it would be related to PureNix.

-   [effect](https://github.com/purescript/effect)

    Nix doesn't have any way of doing arbitrary effects.  In theory, we could
    have an `Effect` type for compatibility, but in practice we wouldn't be
    able to do anything with it.

-   [exceptions](https://github.com/purescript/exceptions)

    Nix doesn't really have exceptions the same way JavaScript does.  Nix does
    have the `builtins.abort`, `assert`, and `builtins.tryEval` functions,
    which may be able to be wrapped up in a nice library.

-   [gen](https://github.com/purescript/gen)

    See the problem with `random`.

-   [minibench](https://github.com/purescript/minibench)

    I don't think Nix has any way of getting the current time (or at least to
    any precision that would be necessary for benchmarking).

-   [parallel](https://github.com/purescript/parallel)

    Nix doesn't have any built-in way of running computations in parallel.

-   [quickcheck](https://github.com/purescript/quickcheck)

    See the problem with `random`.

-   [random](https://github.com/purescript/random)

    As far as I know, Nix doesn't have any sort of random number generator.
    There is no way to generate random numbers.

-   [refs](https://github.com/purescript/refs)

    Nix doesn't have any way of mutating values, so refs are not possible.

-   [tailrec](https://github.com/purescript/tailrec)

    Nix is lazy, so this sort of trampolining isn't necessary.

**Comment by martyall** [comment]
I'm interested in giving one of these a shot, whichever you think is in the intersection of most useful and straight-forward. Any suggestions ?

**Comment by cdepillabout** [comment]
@martyall We recently released a newer version of `purenix` that is based on PureScript-0.15 (instead of 0.14): https://github.com/purenix-org/purenix/pull/51 and https://github.com/purenix-org/purenix/pull/53.

However, we never went back and updated all the actual PureNix PureScript libraries to be based on the versions that are used for the actual PureScript-0.15 package set.  All of our current PureNix libraries are based on PureScript libraries that are used with PureScript-0.14. I'm actually pretty surprised that the [Quick Start instructions](https://github.com/purenix-org/purenix/blob/main/docs/quick-start.md) still work even when you use a 0.15 version of PureNix with the current set of libraries based on PureScript-0.14.

So the most helpful thing would probably be to go through all the PureScript libraries that have already been ported to PureNix, and update their versions to the same versions used in the actual PureScript-0.15 package set.  Although that would likely be quite annoying to do unless you automated or semi-automated the process!

-------------------------------------------------

Other than that, the most helpful thing might be to try to work on some sort of `nix-builtins` library as described in https://github.com/purenix-org/purenix/issues/40.

Or, if you're coming from Haskell, one of the most interesting libraries to try to port to PureNix might be `purescript-transformers`, since that contains all the monad transformers and mtl type-classes.  Having them be available in Nix would be neat.
